crates/oxc_semantic/src/checker/javascript.rs

```diff
@@ -107,9 +107,40 @@ fn invalid_let_declaration(x0: &str, span1: Span) -> OxcDiagnostic {
 
 pub fn check_binding_identifier(ident: &BindingIdentifier, ctx: &SemanticBuilder<'_>) {
     let strict_mode = ctx.strict_mode();
-    // It is a Diagnostic if the StringValue of a BindingIdentifier is "eval" or "arguments" within strict mode code.
+    // In strict mode, `eval` and `arguments` are banned as identifiers.
     if strict_mode && matches!(ident.name.as_str(), "eval" | "arguments") {
-        return ctx.error(unexpected_identifier_assign(&ident.name, ident.span));
+        // `eval` and `arguments` are allowed as the names of declare functions as well as their arguments.
+        //
+        // declare function eval(): void; // OK
+        // declare function arguments(): void; // OK
+        // declare function f(eval: number, arguments: number): number; // OK
+        // declare function g({eval, arguments}: {eval: number, arguments: number}): number; // Error
+        // declare function h([eval, arguments]: [number, number]): number; // Error
+        let is_in_declare_function = |depth: usize| {
+            ctx.nodes
+                .ancestor_kinds(ctx.current_node_id)
+                .nth(depth)
+                .and_then(AstKind::as_function)
+                .is_some_and(|func| matches!(func.r#type, FunctionType::TSDeclareFunction))
+        };
+
+        let is_ok = match ctx.nodes.parent_kind(ctx.current_node_id) {
```

Comment:
It'd be more performant to avoid repeatedly iterating down ancestors from current node.

```rs
let is_declare_function = |kind| {
    kind.as_function()
        .is_some_and(|func| matches!(func.r#type, FunctionType::TSDeclareFunction))
};

let parent_node_id = ctx.nodes.parent_id(ctx.current_node_id);
match ctx.nodes.kind(parent_node_id) {
    AstKind::FormalParameter(_) => {
        let grandparent = ctx.nodes.parent_kind(parent_node_id);
        is_declare_function(grandparent)
    }
    // etc
}
```